src/test/scala/chiselTests/PrintableSpec.scala

```diff
@@ -261,4 +258,115 @@ class PrintableSpec extends AnyFlatSpec with Matchers {
+
+    }
+    generateAndCheck(new MyModule) {
+      case Seq(Printf("F1 = 20.45156 D1 = 10 F1 formatted = 20.45 s1 = 15 l1 = 253", Seq())) =>
+    }.getOrElse(fail())
+  }
+
+  it should "print chisel bits with cf format specifier" in {
+
+    class MyBundle extends Bundle {
+      val foo = UInt(32.W)
+      val bar = UInt(32.W)
+      override def toPrintable: Printable = {
+        cf"Bundle : " +
+          cf"Foo : $foo%x Bar : $bar%x"
+      }
+    }
+    class MyModule extends BasicTester {
+      val b1 = 10.U
+      val w1 = Wire(new MyBundle)
+      w1.foo := 5.U
+      w1.bar := 10.U
+      printf(cf"w1 = $w1")
+    }
+    generateAndCheck(new MyModule) {
+    case Seq(Printf("w1 = Bundle : Foo : %x Bar : %x", Seq("w1.foo", "w1.bar"))) => 
+    }.getOrElse(fail())
+  }
+
+  it should "support names of circuit elements using format specifier including submodule IO with cf format specifier" in {
+    // Submodule IO is a subtle issue because the Chisel element has a different
+    // parent module
+    class MySubModule extends Module {
+      val io = IO(new Bundle {
+        val fizz = UInt(32.W)
+      })
+    }
+    class MyBundle extends Bundle {
+      val foo = UInt(32.W)
+    }
+    class MyModule extends BasicTester {
+      override def desiredName: String = "MyModule"
+      val myWire = Wire(new MyBundle)
+      val myInst = Module(new MySubModule)
+      printf(cf"${myWire.foo}%n")
+      printf(cf"${myWire.foo}%N")
+      printf(cf"${myInst.io.fizz}%N")
+    }
+    generateAndCheck(new MyModule) {
+      case Seq(Printf("foo", Seq()), Printf("myWire.foo", Seq()), Printf("myInst.io.fizz", Seq())) => 
+    }.getOrElse(fail())
+  }
+
+  it should "correctly print strings after modifier" in {
+    class MyModule extends BasicTester {
+      val b1 = 10.U
+      printf(cf"This is here $b1%x!!!! And should print everything else")
+    }
+    generateAndCheck(new MyModule) {
+      case Seq(Printf("This is here %x!!!! And should print everything else",Seq("UInt<4>(\"ha\")"))) => 
+    }.getOrElse(fail())
+  }
+
+  it should "correctly print strings with a lot of literal %% and different format specifiers for Wires" in {
+    class MyModule extends BasicTester {
+      val b1 = 10.U
+      val b2 = 20.U
+      printf(cf"%%  $b1%x%%$b2%b = ${b1%b2}%d %%%% Tail String")
+    }
+
+    generateAndCheck(new MyModule) {
+      case Seq(Printf("%%  %x%%%b = %d %%%% Tail String", Seq(lita,litb,_))) =>
+        assert(lita.contains("UInt<4>") && litb.contains("UInt<5>"))
+    }.getOrElse(fail())
+  }
+
+  it should "not allow unescaped % in the message" in {
+    class MyModule extends BasicTester {
+      printf(cf"This should error out for sure because of % - it should be %%")
+    }
+    a[java.util.UnknownFormatConversionException] should be thrownBy {
+      extractCause[java.util.UnknownFormatConversionException] {
+        ChiselStage.elaborate { new MyModule }
+      }
+    }
+  }
+
```

Comment:
This test is good, it would also be good to have one that only has an unescapted `%` at the end of the `cf"..."` String because this is a common corner case to miss.

Reply:
@jackkoenig  this case is indirectly covered in p"%"  since it calls cf("%%") - but will add one explicitly with cf"%%"

Reply:
> @jackkoenig this case is indirectly covered in p"%" since it calls cf("%%") - but will add one explicitly with cf"%%"

I don't think that's true, since `p"%"` should not error but `cf"...%"` should.

Reply:
Agreed - about cf"%" should error - but cf"%%" should not which is what p"%" translates to. Anyway point taken about this scenario!

Reply:
Just realized I misread your first statement to begin with - my bad!

Reply:
Added the test case to check if single unescaped % does cause an exception.